File: week 1/reachability/reachability.py
def reach(adj, x, y):
    #write your code here
    visited = [0] * len(adj)
    return explore(adj, x, y, visited)
	
def explore(adj, x, y, visited):
    if (x == y):
        return 1
    visited[x] = 1
    for i in xrange(len(adj[x])):
        if (not visited[adj[x][i]]):
            if(explore(adj, adj[x][i], y, visited)):
                return 1
    return 0

if __name__ == '__main__':
    n, m = map(int, raw_input().split())
    adj = [[] for _ in range(n)]
    for i in range(m):
        a, b = map(int, raw_input().split())
        # adjacency list		
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    x, y = map(int, raw_input().split())
    print(reach(adj, x-1, y-1))

File: week 1/reachability/reachability.cpp
#include <iostream>
#include <vector>

using std::vector;
using std::pair;

int explore(vector<vector<int> > &adj, int x, int y, vector<int> &visited) {
	if (x == y) {
		return 1;
	}	
	visited[x] = 1;
	for(int i = 0; i < adj[x].size(); i++) {
		if (!visited[adj[x][i]]) {
			if(explore(adj, adj[x][i], y, visited)) {
				return 1;
			}
		}
	}
	return 0;
}


int reach(vector<vector<int> > &adj, int x, int y) {
  //write your code here
  vector<int> visited(adj.size());
  return explore(adj, x, y, visited);
}

int main() {
  size_t n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
	// adjacency list
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }
  int x, y;
  std::cin >> x >> y;
  std::cout << reach(adj, x - 1, y - 1);
}
